modin/core/execution/ray/common/deferred_execution.py

```diff
@@ -0,0 +1,708 @@
+
+from builtins import NotImplementedError
+from enum import Enum
+from itertools import islice
+from types import GeneratorType
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+import pandas
+import ray
+from ray._private.services import get_node_ip_address
+from ray.util.client.common import ClientObjectRef
+
+ObjectRefType = Union[ray.ObjectRef, ClientObjectRef, None]
+ObjectRefOrListType = Union[ObjectRefType, List[ObjectRefType]]
+ListOrTuple = (list, tuple)
+
+
+class DeferredExecution:
+    """
+    Deferred execution task.
+
+    This class represents a single node in the execution tree. The input is either
+    an object reference or another node on which this node depends.
+    The output is calculated by the specified Callable.
+
+    If the input is a DeferredExecution node, it is executed first and the execution
+    output is used as the input for this one. All the executions are performed in a
+    single batch (i.e. using a single remote call) and the results are saved in all
+    the nodes with ref_counter > 1.
+
+    Parameters
+    ----------
+    data : ObjectRefType or DeferredExecution
+        The execution input.
+    func : callable or ObjectRefType
+        A function to be executed.
+    args : iterable
+        Additional positional arguments to be passed in `func`.
+    kwargs : dict
+        Additional keyword arguments to be passed in `func`.
+    num_returns : int
+        The number of the return values.
+    flat_args : bool, optional
+        True means that there are no lists or DeferredExecution objects in `args`.
+        In this case, no arguments processing is performed and `args` is passed
+        to the remote method as is.
+    flat_kwargs : bool, optional
+        The same as `flat_args` but for the `kwargs` values.
+    """
+
+    def __init__(
+        self,
+        data: Union[
+            ObjectRefType,
+            "DeferredExecution",
+            List[Union[ObjectRefType, "DeferredExecution"]],
+        ],
+        func: Union[Callable, ObjectRefType],
+        args: Iterable[Any],
+        kwargs: Dict[str, Any],
+        num_returns=1,
+        flat_args: Optional[bool] = None,
+        flat_kwargs: Optional[bool] = None,
+    ):
+        ref = DeferredExecution._ref
+        if flat_args is None:
+            if has_list_or_de(args):
+                flat_args = False
+                for a in args:
+                    ref(a)
+            else:
+                flat_args = True
+        if flat_kwargs is None:
+            if has_list_or_de(kwargs.values()):
+                flat_kwargs = False
+                for a in kwargs.values():
+                    ref(a)
+            else:
+                flat_kwargs = True
+        ref(data)
+        self.data = data
+        self.task = (func, args, kwargs, flat_args, flat_kwargs)
```

Comment:
maybe use [`NamedTuple`](https://docs.python.org/3/library/collections.html#collections.namedtuple) here, so the rest of the code would be more readable?
```diff
- self.task[3] and self.task[4]
+ self.task.flat_args and self.task.flat_kwargs
```

Reply:
Tuple is not used any more.